File: GFG/Construct Binary Tree from String with bracket representation/Construct Binary Tree from String with bracket representation.cpp
class Solution{
    
    int ind = 0;
    Node* solve(string str)
    {
        if(str.length() == 0 || ind >= str.length())
           return NULL;
          
        //calculate num
        int num = 0;
        while(ind < str.length() && str[ind] != '(' && str[ind] != ')')
        {
            int val = (int)(str[ind]-'0');
            num = num*10 + val;
            ind++;
        }
        
        //create new root
        Node* root = NULL;
        if(num > 0)
          root = new Node(num);
        
       //for left child
       if(ind < str.length() && str[ind] == '(')
       {
           ind++;
           root->left = solve(str);
       }
        
       if(ind < str.length() && str[ind] == ')')
       {
           ind++;
           return root;
       }
       
       //for right child
       if(ind < str.length() && str[ind] == '(')
       {
           ind++;
           root->right = solve(str);
       }
        
       if(ind < str.length() && str[ind] == ')')
       {
           ind++;
           
       }
        return root;  
        
    }
    
public:
    // function to construct tree from string
    Node *treeFromString(string str){
      
        if(str.length() == 0) return NULL;
        
         return solve(str);
          
    }
};

File: GFG/Construct Binary Tree from String with bracket representation/Construct Binary Tree from String with bracket representation.java
class Solution {
    public static Node treeFromString(String s) {
        // code here
        int[] idx=new int[]{0};
        return dfs(idx,s);
    }
    static Node dfs(int[] idx,String s){
        if(idx[0]>=s.length() || s.charAt(idx[0])==')'){
            idx[0]++;
            return null;
        }
        
        int num=0;
        while(idx[0]<s.length() && Character.isDigit(s.charAt(idx[0]))){
            num=num*10+(s.charAt(idx[0])-'0');
            idx[0]++;
        }
        Node head=new Node(num);
        if(idx[0]<s.length() && s.charAt(idx[0])=='('){
            idx[0]++;
            head.left=dfs(idx,s);
        }
        if(idx[0]<s.length() && s.charAt(idx[0])=='('){
            idx[0]++;
            head.right=dfs(idx,s);
        }
        idx[0]++;
        
        return head;
    }
}
     
